contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol

```diff
@@ -0,0 +1,561 @@
+        uint256 _amountOutputToken      // The amount of output tokens received by the recipient
+    );
+
+    /* ============ Modifiers ============ */
+
+    modifier isWhitelistedIssuanceModule(address _issuanceModule) {
+        require(allowedIssuanceModules[_issuanceModule].isAllowed, "ExchangeIssuance: INVALID ISSUANCE MODULE");
+         _;
+    }
+
+    constructor(
+        address _weth,
+        IController _setController,
+        address[] memory _issuanceModuleAddresses,
+        bool[] memory _issuanceModuleDebtIssuanceFlags,
+        address _swapTarget
+    )
+        public
+    {
+        require(_issuanceModuleAddresses.length == _issuanceModuleDebtIssuanceFlags.length, "ExchangeIssuance: ISSUANCE MODULE ADDRESSES / TYPE FLAGS LENGTH MISMATCH");
+        setController = _setController;
+
+        WETH = _weth;
+        swapTarget = _swapTarget;
+
+        for (uint256 i = 0; i < _issuanceModuleAddresses.length; i++) {
+            _addIssuanceModule(_issuanceModuleAddresses[i], _issuanceModuleDebtIssuanceFlags[i]);
+        }
+
+    }
+
+    /* ============ External Functions ============ */
+
+    /**
+     * Withdraw slippage to selected address
+     *
+     * @param _tokens    Addresses of tokens to withdraw, specifiy ETH_ADDRESS to withdraw ETH
+     * @param _to        Address to send the tokens to
+     */
+    function withdrawTokens(IERC20[] calldata _tokens, address payable _to) external onlyOwner payable {
+        for(uint256 i = 0; i < _tokens.length; i++) {
+            if(address(_tokens[i]) == ETH_ADDRESS){
+                _to.sendValue(address(this).balance);
+            }
+            else{
+                _tokens[i].safeTransfer(_to, _tokens[i].balanceOf(address(this)));
+            }
+        }
+    }
+
+    receive() external payable {
+        // required for weth.withdraw() to work properly
+        require(msg.sender == WETH, "ExchangeIssuance: Direct deposits not allowed");
+    }
+
+    /* ============ Public Functions ============ */
+
+    /**
+     * Whitelists an issuance module
+     *
+     * @param _issuanceModule    Struct containing data on issuance module to add
+     */
+    function addIssuanceModule(address _issuanceModule, bool _isDebtIssuanceModule) public onlyOwner {
+        _addIssuanceModule(_issuanceModule, _isDebtIssuanceModule);
+    }
+
+    /**
+     * Removes an issuance module from the whitelist
+     *
+     * @param _issuanceModuleAddress    Address of issuance module to remove
+     */
+    function removeIssuanceModule(address _issuanceModuleAddress) public onlyOwner {
+        _removeIssuanceModule(_issuanceModuleAddress);
+    }
+
+    /**
+     * Change the _swapTarget
+     *
+     * @param _swapTarget    Address of the swap target contract. (Usually ZeroEx ExchangeProxy)
+     */
+    function setSwapTarget(address _swapTarget) public onlyOwner {
+        swapTarget = _swapTarget;
```

Comment:
0x has been very consistent in the addressing of their exchange proxy. It's "0xdef1c0ded..." for every network (except 1337)

https://github.com/0xProject/protocol/blob/development/packages/contract-addresses/addresses.json

Reply:
Good to know, didn't know this. (Nor do I understand how they did this 😅).

Is this just FYI or do you suggest any changes to our implementation based on this ?

Reply:
I was under the impression that 0x upgraded their swap router from time to time? I thought this would be a nice addition since it would allow us to not have to redeploy whenever 0x had a contract upgrade. If immutability is a big concern I would be open to removing this.

Reply:
The router is a "Diamond" - Proxy so I don't think they regularly deploy that, since they will just redeploy / upgrade the implementation contracts for different feature. The current instance of the router / proxy was deployed June 2020. 

However I agree it might be wasteful to redeploy the whole Exchange Issuance if / when that proxy address changes in the future. 

Also I'd think the risk is manageable since we don't really hold user funds in the EI contract,  only approve the 0x Router with the amount necessary for each trade and use token balances to verify traded amounts. (As opposed to relying on the return value of the swap call).

Reply:
Actually since its a proxy (and unlikely to change), I think we should make it immutable. Allowing it to be changed to an arbitrary address allows the owner to steal funds from any user who has approved the exchange issuance contract as a spender. This can be done by setting an ERC20 address as the router, then doing an exchange issuance with an encoded transferFrom call to rip tokens out of past users addresses. In fact, an attack like this has happened to both DYDX and Gelato recently.

Reply:
Oh!! Wow good eye @ncitron.

Reply:
Great point. I made the swapTarget immutable